Request for Milestone 3 approval: Tic Tac Toe Game

marcoshdezcam opened the pull request:
## This is to request the code review for our project Milestone 3: Game logic. 

Each class is moved into a separate file and the main loop is created on main.rb

### The Authors: 
**Adesoji and Marcos **
[Live preview:](https://repl.it/join/tqsgjwkd-marcoshdezcam1)
In this project, we used; 
### Ruby 
### GitHub/workflows 
Thank You for taking the time to review our code! 💐

Commit:
Correct variable names on Game.play && Working on Board.winner?

Commit:
Change Board.winner? board conditions

Commit:
Remove puts/gets from classes && Create 1 file for each class

Commit:
Fix methods and move game loop to main.rb

Commit:
Remove unnecessary variables

Commit:
Fix rubocop offenses

Commit:
Improve Game.winner? by using Game.winning_moves

majovanilla commented on bin/main.rb:
```diff
@@ -1,109 +1,34 @@
+player2 = Player.new(gets.chomp, 'O')
+TicTacToe = Game.new(player1, player2)
+puts %(Welcome #{TicTacToe.player1.name}! Your token is: #{TicTacToe.player1.token})
+puts %(Welcome #{TicTacToe.player2.name}! Your token is: #{TicTacToe.player2.token})
```
You are repeating the same code for both players. Better to create a method and call it once for each player (it could be stored in the Player class.
TIP: you can return a string instead of using `puts` when you move the code to the Player class.
Remember to validate the player's input and avoid leaving the name blank!

majovanilla commented on bin/main.rb:
```diff
@@ -1,109 +1,34 @@
+puts %(Welcome #{TicTacToe.player1.name}! Your token is: #{TicTacToe.player1.token})
+puts %(Welcome #{TicTacToe.player2.name}! Your token is: #{TicTacToe.player2.token})
+
+loop do
```
Your loop has the same code twice. Better to keep things [DRY](https://deviq.com/don-t-repeat-yourself/) and move the code to a method that you call once for each player.
I would suggest to move the code to the Game class and call it here.
Remember to validate the player's input. No empty inputs, and only 1-9 numbers should be allowed.
Also, you shouldn't allow the player to choose an already taken cell.

AdesojiCodeMaster committed:
Edit draw message

majovanilla commented on bin/main.rb:
```diff
@@ -1,109 +1,34 @@
+  if TicTacToe.winner?(player1) || TicTacToe.draw?(player1)
+    puts %(Congratulations #{player1.name}, you won!) if TicTacToe.winner?(player1)
+    puts %(It's a draw!) if TicTacToe.draw?(player1)
+    break
```
You should create a method to store this code. 
I suggest having one for the winner and one for the draw,

TIP: you don't need to check for a winner until the **fifth** move. You only need to check for a draw on the **ninth** move if there's no winner.

majovanilla commented on lib/board.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative '../lib/player.rb'
+
+class Board
+  attr_accessor :game_board, :slots_taken
```
Remember that having `attr_accessor` allows your variable to be read and written from anywhere in your application. Make sure you actually need them or change them to `attr_reader` whenever you can.

majovanilla commented on lib/player.rb:
```diff
@@ -0,0 +1,10 @@
+  attr_accessor :name, :next_move, :moves, :token
+
+  def initialize(name, token)
+    @name = name
+    @next_move = []
+    @moves = []
+    @token = token
```
Do you really need to store the player's moves?
What happens if you just mark the moves on the board and check for the player's mark in the winning positions?
That way you can simplify your `winner?` and `mark_board` methods.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,31 @@
+    @board_tic.slots_taken << player.next_move && player.moves << player.next_move
+  end
+
+  def winner?(player)
+    winning_moves.any?(player.moves)
+  end
```
If you store the board state into an array, you can use the `all_equal` method to check if there is a winner.
It doesn't work on some horizontal wins.
![Screenshot from 2020-08-26 19-32-43](https://user-images.githubusercontent.com/51926937/91370080-02d74700-e7d3-11ea-9b7b-65b6a1d339d1.png)

majovanilla commented on lib/player.rb:
```diff
@@ -0,0 +1,10 @@
+class Player
+  attr_accessor :name, :next_move, :moves, :token
```
Are you sure you need to allow writing and reading permissions to all the variables?

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def mark_board(player)
+    return if @board_tic.marked?(player)
```
This is not the best way to avoid selecting the same cell. I can mark select the same cell any time I want but I get no warning.
Check how you are switching turns. I can type anything (even if it's wrong) and my turn ends instead of giving me notice and allowing me to select again.

majovanilla commented on lib/game.rb:
```diff
@@ -0,0 +1,31 @@
+    winning_moves.any?(player.moves)
+  end
+
+  def draw?(player)
+    @board_tic.slots_taken.size == 9 && !winner?(player)
+  end
```
This method is not working all the time. It marks a draw before the game is over. That happens when I choose the same cell multiple times.
![Screenshot from 2020-08-26 19-30-44](https://user-images.githubusercontent.com/51926937/91369969-bb50bb00-e7d2-11ea-9111-645e9603fdd9.png)

majovanilla reviewed (changes requested):
# Changes required :mag: 

Hey team, 👋

## Good job on this project! 👍

- Good Readme file
- Good Pull Request Message

### There are still some changes 👇 you need to implement to approve this request 🔍

### Required Changes: ✍️

- Please validate all player's inputs. I left the comments below.
- I also suggested a way to improve your method to check for a winner.
- You need to double-check your win and draw methods because they don't work some times.
- Don't disable the linters rules. Instead, fix the code.
- Check the rest of the comments I left.

## Please, make all REQUIRED changes and submit your request again! 😀

If you have any questions, feel free to send me a message through Slack (@majo)
I would be happy to help! 👩‍💻 


Happy coding 🤓

--Maria Reyes

AdesojiCodeMaster committed:
Reset the attr_reader and attr_accessor

Commit:
TSA required: Move welcome message to Game.welcome

Commit:
Change to attr_accesor for Player.next_move

Commit:
Add validate_name and create_token to Player. Main validates input

Commit:
Delete Player.validate_token. && Add Game.validate_players to DRY main

Commit:
Validate number introduced by user before marking board

AdesojiCodeMaster committed:
Adjust game validation method

Commit:
Create Game.ask_moves && Move mark_board? to board.rb

Commit:
Fix Game.winner? to iterate on players

Commit:
Change Game.winner? to previous solution

AdesojiCodeMaster committed:
work on winning method

Commit:
Improve Board.show_board using each

Commit:
TSA required: Fix winner? and draw?. Game works!

Check for winner after the fifth move
Check for draw until the game has finished

Commit:
Refactor Game and fix linter errors. Everything works!

marcoshdezcam commented:
## Request for code review after implementing TSA required changes
### Milestone 2: 
All the features of the game are working and the issues with the winning patterns are fixed.
The winner is now selected based on the marks of the board using all? And, not on winning moves. 
Functions refactored to DRY code

### The Authors: 
**Adesoji and Marcos **
[Live preview:](https://repl.it/join/gjrzjjbm-marcoshdezcam1)

### Made using:  
#### - Ruby
#### - Gitflow Workflow

Thank you for taking the time to review our code! 💐

Commit:
Update live preview on README.md

adaorachi commented on bin/main.rb:
```diff
@@ -1,109 +1,6 @@
-  attr_accessor :tic_tac_toe, :player1, :player2
+require_relative '../lib/game.rb'
 
-  def initialize
-    @tic_tac_toe = Board.new
-    puts %(First player; please type your name!)
-    @player1 = Player.new(gets.chomp, 'X')
-    puts %(Second player; please type your name!)
-    @player2 = Player.new(gets.chomp, 'O')
-  end
-
-  def welcome
-    puts %(Welcome on board #{@player1.name} and #{@player2.name}, let the Tic-Tac-Toe game begins!)
-    puts %(Player 1: #{@player1.name} has token: #{@player1.token})
-    puts %(Player 2: #{@player2.name} has token: #{@player2.token})
-  end
-
```
- [x] Kindly remove all `puts` statements here. You should handle the `puts` and user interaction in the main.rb file. One of the possible ways to go about this is to make your methods accept one or two parameters and when you call the methods in the main.rb, it takes in the user's input as arguments and with that, you can use the value to compute your logic and then print your desired string based on the input from the user. 

   - With this, you can get to keep your code DRY and not repeating them as you have in this code.

Another way possible is to return strings instead of printing them just like the previous reviewer @majovanilla  mentioned.

adaorachi commented on bin/main.rb:
```diff
@@ -1,109 +1,6 @@
-    @game_board.each { |itr| puts %( | #{itr[0]} | | #{itr[1]} | | #{itr[2]} |) }
-  end
-
-  def ask_move(player)
-    puts %(#{player.name}'s turn. Use a number between 1 and 9: )
-    player.next_move = gets.chomp.to_i
-    while marked?(player)
-      puts %(Slot already taken! Choose a different number.)
-      player.next_move = gets.chomp.to_i
-    end
-    until player.next_move.between?(1, 9)
-      puts %(Invalid move. Only use numbres bwteen 1 and 9. Try again: )
-      player.next_move = gets.chomp.to_i
-    end
-    @slots_taken << player.next_move
-  end
```
- [x] Do ensure you remove all `puts` statements here. Kindly see my suggestion above. 
#### **Note: You used the puts statements in other several places in this file and in the board.rb file. Kindly remove all instances of it and put them in your main.rb file.**

adaorachi commented on lib/game.rb:
```diff
@@ -0,0 +1,104 @@
+require_relative '../lib/board.rb'
+class Game
+  attr_reader :board, :players, :winner, :draw
+
```
- [x] You do not need these attr_readers here. There are not read outside this file and you already have them as class variables in the initialize method.

adaorachi reviewed (changes requested):
### Hi @marcoshdezcam and @AdesojiCodeMaster, :wave: 
You have done an awesome job on implementing the changes requested in the previous review and I really commend that. 
You also did a good job covering the edges cases mentioned too.
However, there is still a couple of changes to implement in order to get your project approved.
So hang in there, you are just a step away from completing this project.

---

### STATUS: CHANGES REQUIRED :recycle: 

---

- [x] Kindly ensure that all `puts` or `chomp` statements are used only in the main.rb file. Currently, you have a handful of them in the logic files. Here is a [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from pathwright for clarification.
   - I noticed you moved all game flow implementation from your main.rb file into the logic files and only returned two methods in the main.rb. 
I'm aware you are trying to keep your code DRY (which is very important) but at the same time, you should also remember that the project requirements say that the `main.rb` file should be responsible for user interaction (use of `puts` and `chomp` statements) and which consists of the game flow implementation too. 
With this, you can keep your code DRY without using any `puts` or `chomp` statement in the logic files

- [x] Your game should be able to display a message when the game is a tie. Currently, it does not. Kindly ensure you fix the issue. 

- [x] See the other changes I commented below.

After implementing the changes, you can request for another code review.
---

Should you have any question, do reach out to me on Slack with the username @adaorachi

Happy Coding :relaxed: 

MaryAnn Chukwuka.

Commit:
TSE required: All puts/gets moved to main.rb

All user interaction and game flow moved to main.rb as per project requirements
and TSE suggestions.

marcoshdezcam commented:
## Request for code review after implementing TSA required changes
### Milestone 2: 
All the puts/gets statements and the game flow of the game have been moved to the main.rb file as per project requirements and TSE suggestions. 

### The Authors: 
**Adesoji and Marcos **
[Live preview:](https://repl.it/join/gjrzjjbm-marcoshdezcam1)

### Made using:  
#### - Ruby
#### - Gitflow Workflow

Thank you for taking the time to review our code! 💐

voscarmv commented on lib/board.rb:
```diff
@@ -0,0 +1,31 @@
+
+  def show_board
+    @game_board.each do |row|
+      row.each { |col| print %(| #{col} | ) }
+      puts %()
```
From Pathwright:
![image](https://user-images.githubusercontent.com/2739245/91670026-4c7ba680-eadf-11ea-9b74-0348e4c57c4f.png)

- Please avoid `puts` and `print` everywhere except in `main.rb`. To achieve this, you may return string values from your `Board` class, and then use `puts` in `main.tb` to display the returned value.

voscarmv reviewed (changes requested):
# Impressive work, team! 🖥️ 👀 ✨ 

- Excellent work implementing the changes required by the previous TSE.

## However, some changes are still required... 🧰 🔨 🔧 

- You still have some `print` and `puts` calls outside of `main.rb`. Please move those to `main.rb`! (See code comments for details)

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

Commit:
Move Board.show_board to main.rb as requested by TSE

marcoshdezcam commented:
## Request for code review after implementing TSE required changes
### Milestone 2: 
Move Board.show_board to main.rb as requested by TSE. 

### The Authors: 
**Adesoji and Marcos **
[Live preview:](https://repl.it/join/gjrzjjbm-marcoshdezcam1)

### Made using:  
#### - Ruby
#### - Gitflow Workflow

Thank you for taking the time to review our code! 💐

maelfosso reviewed (approved):
# Congratulations :clap: 

Thank you for applying the suggested changes.
You did a really good job. You can merge it now :+1: 

Good continuation :handshake: 

Mael FOSSO
